File: typescript/packages/x402/src/client/x402Client.ts
/**
 * X402Client - Core payment client for x402 protocol
 * 
 * Manages payment mechanism registry and coordinates payment flows.
 */

import type {
  PaymentRequirements,
  PaymentPayload,
  PaymentPermitContext,
} from '../types/index.js';
import { DefaultTokenSelectionStrategy } from './tokenSelection.js';
import type { TokenSelectionStrategy } from './tokenSelection.js';

/** Client mechanism interface */
export interface ClientMechanism {
  /** Get payment scheme name */
  scheme(): string;
  
  /** Create payment payload */
  createPaymentPayload(
    requirements: PaymentRequirements,
    resource: string,
    extensions?: { paymentPermitContext?: PaymentPermitContext }
  ): Promise<PaymentPayload>;
}

/** Client signer interface */
export interface ClientSigner {
  /** Get signer address */
  getAddress(): string;
  
  /** Sign raw message */
  signMessage(message: Uint8Array): Promise<string>;
  
  /** Sign typed data (EIP-712) */
  signTypedData(
    domain: Record<string, unknown>,
    types: Record<string, unknown>,
    message: Record<string, unknown>
  ): Promise<string>;
  
  /** Check token allowance */
  checkAllowance(token: string, amount: bigint, network: string): Promise<bigint>;
  
  /** Ensure sufficient allowance */
  ensureAllowance(
    token: string,
    amount: bigint,
    network: string,
    mode?: 'auto' | 'interactive' | 'skip'
  ): Promise<boolean>;
}

/** Payment requirements selector function */
export type PaymentRequirementsSelector = (
  requirements: PaymentRequirements[]
) => PaymentRequirements;

/** Filter options for selecting payment requirements */
export interface PaymentRequirementsFilter {
  scheme?: string;
  network?: string;
  maxAmount?: string;
}

/** Registered mechanism entry */
interface MechanismEntry {
  pattern: string;
  mechanism: ClientMechanism;
  priority: number;
}

/**
 * X402Client - Core payment client
 * 
 * Manages payment mechanisms and coordinates payment flows.
 */
export class X402Client {
  private mechanisms: MechanismEntry[] = [];
  private tokenStrategy?: TokenSelectionStrategy;

  constructor(options?: { tokenStrategy?: TokenSelectionStrategy }) {
    this.tokenStrategy = options?.tokenStrategy;
  }

  /**
   * Register payment mechanism for network pattern
   * 
   * @param networkPattern - Network pattern (e.g. "eip155:*", "tron:shasta")
   * @param mechanism - Payment mechanism instance
   * @returns this for method chaining
   */
  register(networkPattern: string, mechanism: ClientMechanism): X402Client {
    const priority = this.calculatePriority(networkPattern);
    this.mechanisms.push({
      pattern: networkPattern,
      mechanism,
      priority,
    });
    this.mechanisms.sort((a, b) => b.priority - a.priority);
    return this;
  }

  /**
   * Select payment requirements from available options
   * 
   * @param accepts - Available payment requirements
   * @param filters - Optional filters
   * @returns Selected payment requirements
   */
  selectPaymentRequirements(
    accepts: PaymentRequirements[],
    filters?: PaymentRequirementsFilter
  ): PaymentRequirements {
    let candidates = accepts;

    if (filters?.scheme) {
      candidates = candidates.filter(r => r.scheme === filters.scheme);
    }

    if (filters?.network) {
      candidates = candidates.filter(r => r.network === filters.network);
    }

    if (filters?.maxAmount) {
      const max = BigInt(filters.maxAmount);
      candidates = candidates.filter(r => BigInt(r.amount) <= max);
    }

    candidates = candidates.filter(r => this.findMechanism(r.network) !== null);

    if (candidates.length === 0) {
      throw new Error('No supported payment requirements found');
    }

    if (this.tokenStrategy) {
      return this.tokenStrategy.select(candidates);
    }

    return new DefaultTokenSelectionStrategy().select(candidates);
  }

  /**
   * Create payment payload for given requirements
   * 
   * @param requirements - Selected payment requirements
   * @param resource - Resource URL
   * @param extensions - Optional extensions
   * @returns Payment payload
   */
  async createPaymentPayload(
    requirements: PaymentRequirements,
    resource: string,
    extensions?: { paymentPermitContext?: PaymentPermitContext }
  ): Promise<PaymentPayload> {
    const mechanism = this.findMechanism(requirements.network);
    if (!mechanism) {
      throw new Error(`No mechanism registered for network: ${requirements.network}`);
    }

    return mechanism.createPaymentPayload(requirements, resource, extensions);
  }

  /**
   * Handle payment required response
   * 
   * @param accepts - Available payment requirements
   * @param resource - Resource URL
   * @param extensions - Optional extensions
   * @param selector - Optional custom selector
   * @returns Payment payload
   */
  async handlePayment(
    accepts: PaymentRequirements[],
    resource: string,
    extensions?: { paymentPermitContext?: PaymentPermitContext },
    selector?: PaymentRequirementsSelector
  ): Promise<PaymentPayload> {
    const requirements = selector
      ? selector(accepts)
      : this.selectPaymentRequirements(accepts);

    return this.createPaymentPayload(requirements, resource, extensions);
  }

  /**
   * Find mechanism for network
   */
  private findMechanism(network: string): ClientMechanism | null {
    for (const entry of this.mechanisms) {
      if (this.matchPattern(entry.pattern, network)) {
        return entry.mechanism;
      }
    }
    return null;
  }

  /**
   * Match network with pattern
   */
  private matchPattern(pattern: string, network: string): boolean {
    if (pattern === network) return true;
    if (pattern.endsWith(':*')) {
      const prefix = pattern.slice(0, -1);
      return network.startsWith(prefix);
    }
    return false;
  }

  /**
   * Calculate priority for pattern (more specific = higher priority)
   */
  private calculatePriority(pattern: string): number {
    if (pattern.endsWith(':*')) return 1;
    return 10;
  }
}

File: python/x402/src/bankofai/x402/clients/x402_client.py
"""
X402Client - Core payment client for x402 protocol
"""

import logging
from typing import TYPE_CHECKING, Any, Callable, Protocol

from bankofai.x402.exceptions import UnsupportedNetworkError
from bankofai.x402.types import (
    PaymentPayload,
    PaymentRequirements,
)

if TYPE_CHECKING:
    from bankofai.x402.clients.token_selection import TokenSelectionStrategy

logger = logging.getLogger(__name__)


class ClientMechanism(Protocol):
    """Client mechanism interface"""

    def scheme(self) -> str:
        """Get the payment scheme name"""
        ...

    def get_signer(self) -> Any:
        """Return the signer used by this mechanism, or None."""
        ...

    async def create_payment_payload(
        self,
        requirements: PaymentRequirements,
        resource: str,
        extensions: dict[str, Any] | None = None,
    ) -> PaymentPayload:
        """Create payment payload"""
        ...


PaymentRequirementsSelector = Callable[[list[PaymentRequirements]], PaymentRequirements]


class PaymentPolicy(Protocol):
    """Policy that filters or reorders payment requirements.

    Policies are applied in order after mechanism filtering and before
    token selection. Return a subset (or reordered list) of the input.
    """

    async def apply(
        self,
        requirements: list[PaymentRequirements],
    ) -> list[PaymentRequirements]:
        """Apply this policy to the given requirements."""
        ...


class PaymentRequirementsFilter:
    """Filter options for selecting payment requirements"""

    def __init__(
        self,
        scheme: str | None = None,
        network: str | None = None,
    ):
        self.scheme = scheme
        self.network = network


class MechanismEntry:
    """Registered mechanism entry"""

    def __init__(self, pattern: str, mechanism: ClientMechanism, priority: int):
        self.pattern = pattern
        self.mechanism = mechanism
        self.priority = priority


class X402Client:
    """
    Core payment client for x402 protocol.

    Manages payment mechanism registry and coordinates payment flow.
    """

    def __init__(
        self,
        token_strategy: "TokenSelectionStrategy | None" = None,
    ) -> None:
        """
        Initialize X402Client.

        Args:
            token_strategy: Strategy for selecting which token to pay with.
                            If None, uses first available option.
        """
        self._mechanisms: list[MechanismEntry] = []
        self._policies: list[PaymentPolicy] = []
        self._token_strategy = token_strategy

    def register_policy(self, policy: PaymentPolicy) -> "X402Client":
        """
        Register a payment policy.

        Policies are applied in order after mechanism filtering
        and before token selection.

        Args:
            policy: Callable that filters/reorders payment requirements

        Returns:
            self for method chaining
        """
        self._policies.append(policy)
        return self

    def register(self, network_pattern: str, mechanism: ClientMechanism) -> "X402Client":
        """
        Register a payment mechanism for a network pattern.

        Args:
            network_pattern: Network pattern (e.g., "eip155:*", "tron:shasta")
            mechanism: Payment mechanism instance

        Returns:
            self for method chaining
        """
        priority = self._calculate_priority(network_pattern)
        logger.info(
            f"Registering mechanism for pattern '{network_pattern}' with priority {priority}"
        )
        self._mechanisms.append(MechanismEntry(network_pattern, mechanism, priority))
        self._mechanisms.sort(key=lambda e: e.priority, reverse=True)
        return self

    async def select_payment_requirements(
        self,
        accepts: list[PaymentRequirements],
        filters: PaymentRequirementsFilter | None = None,
    ) -> PaymentRequirements:
        """
        Select payment requirements from available options.

        Applies filters, then delegates to the configured token selection
        strategy. Falls back to DefaultTokenSelectionStrategy if none is set.

        Args:
            accepts: Available payment requirements
            filters: Optional filters

        Returns:
            Selected payment requirements

        Raises:
            ValueError: No supported payment requirements found
        """
        logger.info(f"Selecting payment requirements from {len(accepts)} options")
        logger.debug(f"Available payment requirements: {[r.model_dump() for r in accepts]}")

        candidates = list(accepts)

        if filters:
            if hasattr(filters, "scheme") and filters.scheme:
                candidates = [r for r in candidates if r.scheme == filters.scheme]
                logger.debug(f"After scheme filter: {len(candidates)} candidates")
            if hasattr(filters, "network") and filters.network:
                candidates = [r for r in candidates if r.network == filters.network]
                logger.debug(f"After network filter: {len(candidates)} candidates")

        candidates = [
            r for r in candidates if self._find_mechanism(r.scheme, r.network) is not None
        ]
        logger.debug(f"After mechanism filter: {len(candidates)} candidates")

        for policy in self._policies:
            candidates = await policy.apply(candidates)
            logger.debug(f"After policy: {len(candidates)} candidates")

        if not candidates:
            logger.error("No supported payment requirements found")
            raise UnsupportedNetworkError("No supported payment requirements found")

        if self._token_strategy:
            selected = await self._token_strategy.select(candidates)
        else:
            from bankofai.x402.clients.token_selection import DefaultTokenSelectionStrategy

            selected = await DefaultTokenSelectionStrategy().select(candidates)

        logger.info(
            "Selected payment requirement: network=%s, scheme=%s, amount=%s",
            selected.network,
            selected.scheme,
            selected.amount,
        )
        return selected

    async def create_payment_payload(
        self,
        requirements: PaymentRequirements,
        resource: str,
        extensions: dict[str, Any] | None = None,
    ) -> PaymentPayload:
        """
        Create payment payload for given requirements.

        Args:
            requirements: Selected payment requirements
            resource: Resource URL
            extensions: Optional extensions

        Returns:
            Payment payload
        """
        logger.info(
            f"Creating payment payload for scheme={requirements.scheme}, "
            f"network={requirements.network}, resource={resource}"
        )
        mechanism = self._find_mechanism(requirements.scheme, requirements.network)
        if mechanism is None:
            logger.error(
                f"No mechanism registered for scheme={requirements.scheme}, "
                f"network={requirements.network}"
            )
            raise UnsupportedNetworkError(
                f"No mechanism registered for scheme={requirements.scheme}, "
                f"network={requirements.network}"
            )

        logger.debug(f"Using mechanism: {mechanism.__class__.__name__}")
        payload = await mechanism.create_payment_payload(requirements, resource, extensions)
        logger.info("Payment payload created successfully")
        return payload

    async def handle_payment(
        self,
        accepts: list[PaymentRequirements],
        resource: str,
        extensions: dict[str, Any] | None = None,
        selector: PaymentRequirementsSelector | None = None,
    ) -> PaymentPayload:
        """
        Handle payment required response.

        Args:
            accepts: Available payment requirements
            resource: Resource URL
            extensions: Optional extensions
            selector: Optional custom selector

        Returns:
            Payment payload
        """
        if selector:
            requirements = selector(accepts)
        else:
            requirements = await self.select_payment_requirements(accepts)

        return await self.create_payment_payload(requirements, resource, extensions)

    def _find_mechanism(self, scheme: str, network: str) -> ClientMechanism | None:
        """Find mechanism for scheme and network"""
        for entry in self._mechanisms:
            if entry.mechanism.scheme() == scheme and self._match_pattern(entry.pattern, network):
                return entry.mechanism
        return None

    def _match_pattern(self, pattern: str, network: str) -> bool:
        """Match network against pattern"""
        if pattern == network:
            return True
        if pattern.endswith(":*"):
            prefix = pattern[:-1]
            return network.startswith(prefix)
        return False

    def _calculate_priority(self, pattern: str) -> int:
        """Calculate priority for pattern (more specific = higher priority)"""
        if pattern.endswith(":*"):
            return 1
        return 10
